libbfddp/bfddp_extra.c

```diff
@@ -0,0 +1,858 @@
+}
+
+static void
+bfddp_session_update_dummy(__attribute__((unused)) struct bfd_session *bs,
+			   __attribute__((unused)) void *arg)
+{
+}
+
+static void
+bfddp_session_state_change_dummy(
+	__attribute__((unused)) struct bfd_session *bs,
+	__attribute__((unused)) void *arg,
+	__attribute__((unused)) enum bfd_state_value ostate,
+	__attribute__((unused)) enum bfd_state_value nstate)
+{
+}
+
+/* Our selected BFD integration callbacks. */
+struct bfddp_callbacks bfddp_callbacks;
+
+void
+bfddp_initialize(struct bfddp_callbacks *bc)
+{
+	bfddp_callbacks = *bc;
+
+	if (bfddp_callbacks.bc_session_new == NULL)
+		bfddp_callbacks.bc_session_new = bfddp_session_new_dummy;
+	if (bfddp_callbacks.bc_session_update == NULL)
+		bfddp_callbacks.bc_session_update = bfddp_session_update_dummy;
+	if (bfddp_callbacks.bc_session_free == NULL)
+		bfddp_callbacks.bc_session_free = bfddp_session_update_dummy;
+	if (bfddp_callbacks.bc_state_change == NULL)
+		bfddp_callbacks.bc_state_change =
+			bfddp_session_state_change_dummy;
+
+#define CALLBACK_CHECK(cb)                                                     \
+	do {                                                                   \
+		if (bfddp_callbacks.cb == NULL) {                              \
+			fprintf(stderr, "%s: callback " #cb " not set\n",      \
+				__func__);                                     \
+			exit(1);                                               \
+		}                                                              \
+	} while (0)
+
+	CALLBACK_CHECK(bc_tx_control);
+	CALLBACK_CHECK(bc_tx_control_update);
+	CALLBACK_CHECK(bc_tx_control_stop);
+	CALLBACK_CHECK(bc_rx_control_update);
+	CALLBACK_CHECK(bc_rx_control_stop);
+
+#undef CALLBACK_CHECK
+}
+
+/*
+ * Session functions.
+ */
+static void
+bfddp_session_set_slowstart(struct bfd_session *bs)
+{
+	/* Slow start settings. */
+	bs->bs_cur_dmultiplier = SLOWSTART_DMULT;
+	bs->bs_cur_tx = SLOWSTART_TX;
+	bs->bs_cur_rx = SLOWSTART_RX;
+	bs->bs_cur_erx = SLOWSTART_ERX;
+}
+
+static void
+bfddp_session_reset_remote(struct bfd_session *bs)
+{
+	/* Default remote settings. */
+	bs->bs_rdmultiplier = SLOWSTART_DMULT;
+	bs->bs_rtx = SLOWSTART_TX;
+	bs->bs_rrx = SLOWSTART_RX;
+	bs->bs_rerx = SLOWSTART_ERX;
+}
+
+void
+bfddp_session_update(struct bfd_session *bs, void *arg,
+		     const struct bfddp_session *bds)
+{
+	bool timers_changed = false;
+	struct in_addr *ia;
+	uint16_t port;
+	uint32_t min_rx, min_tx;
+	uint32_t flags = ntohl(bds->flags);
+
+	/*
+	 * Load flags.
+	 *
+	 * NOTE: normalize boolean values (e.g. `!!`) so packet build functions
+	 * can use it with shift (e.g. (multihop << X)).
+	 */
+	bs->bs_admin_shutdown = !!(flags & SESSION_SHUTDOWN);
+	bs->bs_multihop = !!(flags & SESSION_MULTIHOP);
+	bs->bs_passive = !!(flags & SESSION_PASSIVE);
+	bs->bs_demand = !!(flags & SESSION_DEMAND);
+	bs->bs_cbit = !!(flags & SESSION_CBIT);
+	bs->bs_echo = !!(flags & SESSION_ECHO);
+
+	if (bs->bs_multihop)
+		port = htons(BFD_MULTI_HOP_PORT);
```

Comment:
(they are symmetric, so it doesn't matter but..) here conversion is Host -> Network, and 20 lines before it was Network->Host